### PR TITLE
#2123: split up auth & data providers

### PR DESCRIPTION
**Pull request by samrichca**
**Description of changes**

- split up authprovider into auth & data providers
- add ego URLs to config
- remove duplicate code from /logged-in
- added types to _app.tsx
- install jest, add some unit tests

**Type of Change**

- [ ] Bug
- [ ] Styling
- [x] New Feature

**Checklist before requesting review:**

- Design (select one):
  - [ ] Matches design:
    - component sizes, spacing, and styles
    - font size, weight, colour
    - spelling has been double checked
  - [ ] No design provided, screenshot of changes included in PR
  - [x] No changes to UI
- UI Kit (select one):
  - [ ] New Components with storybook stories created
  - [ ] Modified Existing components and updates stories
  - [x] No new UIKit components or changes
- [x] Feature is minimally responsive
- [x] Manual testing of UI feature
- [x] Add copyrights to new files
- [x] Connected ticket to PR

### CODE REVIEW

File: components/pages/file-repository/hooks/useDebounce.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import React, { useState, useEffect } from 'react';
 
 // from https://usehooks.com/useDebounce/
 
-export default <T>(value: T, delay: number): T => {
+const useDebounce = <T>(value: T, delay: number): T => {
```
Comment (samrichca) [inline components/pages/file-repository/hooks/useDebounce.ts:5]: this was giving me a "name your function" warning repeatedly so i named it

File: global/config.ts [hunk 2]
```diff
@@ -22,19 +22,21 @@ import getNextConfig from 'next/config';
...
     EGO_PUBLIC_KEY:
       publicConfig.EGO_PUBLIC_KEY ||
       `-----BEGIN PUBLIC KEY-----\r\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA0lOqMuPLCVusc6szklNXQL1FHhSkEgR7An+8BllBqTsRHM4bRYosseGFCbYPn8r8FsWuMDtxp0CwTyMQR2PCbJ740DdpbE1KC6jAfZxqcBete7gP0tooJtbvnA6X4vNpG4ukhtUoN9DzNOO0eqMU0Rgyy5HjERdYEWkwTNB30i9I+nHFOSj4MGLBSxNlnuo3keeomCRgtimCx+L/K3HNo0QHTG1J7RzLVAchfQT0lu3pUJ8kB+UM6/6NG+fVyysJyRZ9gadsr4gvHHckw8oUBp2tHvqBEkEdY+rt1Mf5jppt7JUV7HAPLB/qR5jhALY2FX/8MN+lPLmb/nLQQichVQIDAQAB\r\n-----END PUBLIC KEY-----`,
     AUTH_DISABLED: publicConfig.AUTH_DISABLED === 'true',
     GA_TRACKING_ID: publicConfig.GA_TRACKING_ID || '',
-    EGO_URL: urlJoin(
-      publicConfig.EGO_API_ROOT || '',
-      `/api/oauth/login/google?client_id=${publicConfig.EGO_CLIENT_ID || ''}`,
-    ),
+    EGO_URL: urlJoin(egoApiRoot, `/api/oauth/login/google?client_id=${egoClientId}`),
```
Comment (samrichca) [inline global/config.ts:36]: put all ego URLs in config to keep track of them better

File: pages/logged-in.tsx [hunk 3]
```diff
@@ -17,72 +17,15 @@
...
-import { css } from 'uikit';
-import DnaLoader from 'uikit/DnaLoader';
-import useTheme from 'uikit/utils/useTheme';
-import urlJoin from 'url-join';
-import { getPermissionsFromToken } from 'global/utils/egoJwt';
-import { useRouter } from 'next/router';
+import DefaultLayout from '../components/pages/DefaultLayout';
+import FullPageLoader from '../components/placeholders/FullPageLoader';
 
 export default createPage({ isPublic: true })(() => {
```
Comment (samrichca) [inline pages/logged-in.tsx:25]: this page had repeated logic and components from app.tsx so I removed the redundancies.

File: pages/_app.tsx [hunk 4]
```diff
@@ -189,33 +182,33 @@ class Root extends App<
...
-      initialPermissions,
       maintenanceModeOn: MAINTENANCE_MODE_ON,
     };
   }
 
-  isInOauthMode = (props) => {
-    if (get(props.ctx.query, 'redirect')) {
-      const parsed = queryString.parseUrl(decodeURIComponent(props.ctx.query.redirect));
+  isInOauthMode = (props: CustomAppProps) => {
+    if (props.ctx.asPath === LOGGED_IN_PATH) {
```
Comment (samrichca) [inline pages/_app.tsx:190]: this is handling logic removed from the /logged-in page

File: pages/_app.tsx [hunk 5]
```diff
@@ -60,36 +76,14 @@ const getRedirect = (ctxAsPath: string | undefined): string =>
 
 const enforceLogin = ({ ctx }: { ctx: NextPageContext }) => {
   const loginRedirect = getRedirect(ctx.asPath);
-  redirect(ctx.res, loginRedirect);
-};
-
-type RootGetInitialPropsData = {
```
Comment (samrichca) [inline pages/_app.tsx:66]: this wasn't being used

File: pages/_app.tsx [hunk 6]
```diff
@@ -259,13 +252,20 @@ class Root extends App<
...
+          if (redirect) {
+            const redirectFromURL = decodeURIComponent(redirect as string);
+            const obj = queryString.parseUrl(redirectFromURL || '');
+            const target = queryString.stringifyUrl({
+              ...obj,
+              query: omit(obj.query, OAUTH_QUERY_PARAM_NAME),
+            });
+            location.assign(target);
+          } else {
+            const redirectFromPermissions = getDefaultRedirectPathForUser(
```
Comment (samrichca) [inline pages/_app.tsx:264]: this is what was removed from /logged-in

**Review by mistryrn — APPROVED** [review]
Awesome work here. Lots of nice small refactors included as well to clean things up! Only thing I noticed is that the copyright headers on the new files are all still `2020`, might want to bump those up to `2022`?

**Review by mistryrn — APPROVED** [review]
New changes look good! 👍

**Pull request merged**